internal/evaluator/builder.go

```diff
@@ -81,29 +79,36 @@ func (bldr *Builder) load(desc protoreflect.MessageDescriptor) MessageEvaluator
 // descriptor, or lazily constructs a new one. This method is thread-safe via
 // locking.
 func (bldr *Builder) loadOrBuild(desc protoreflect.MessageDescriptor) MessageEvaluator {
-	bldr.mtx.RLock()
-	if eval, ok := bldr.cache[desc]; ok {
-		bldr.mtx.RUnlock()
+	cache := *bldr.cache.Load()
+	if eval, ok := cache[desc]; ok {
 		return eval
 	}
-	bldr.mtx.RUnlock()
-
-	bldr.mtx.Lock()
-	defer bldr.mtx.Unlock()
-	return bldr.build(desc)
+	newCache := make(map[protoreflect.MessageDescriptor]*message, len(cache)+1)
+	for k, v := range cache {
+		newCache[k] = v
+	}
+	msgEval := bldr.build(desc, newCache)
+	bldr.cache.Store(&newCache)
+	return msgEval
```

Comment:
Would two concurrent calls to `loadOrBuild` potentially compete to get their evaluator into the cache with this?

```
A ---> Load Cache: C ---> Copy Cache: Ca ---> Cache New Eval ---> Store Cache: Ca \
                                                                                   >-- C?
B ---> Load Cache: C ---> Copy Cache: Cb ---> Cache New Eval ---> Store Cache: Cb /
```

Would using CAS help in this case?

Reply:
Yep. A or B, one wins, next call might have to redo the work. Eventually the cache will be consistent. I've added back the lock to serialize writes. What do you mean by CAS?

Reply:
[CompareAndSwap](https://pkg.go.dev/sync/atomic#Pointer.CompareAndSwap)

1. On load, hold onto the returned `prevPtr`
2. Make copy of the underlying Cache 
3. Build as before populating the evaluators into the Cache
4. Instead of calling `Store`, call `CompareAndSwap` using `prevPtr`
5. If successful, we're good to go. Done.
6. If it's not, it means there's a new value in the atomic.
7. Load again into `prevPtr`. 
8. Copy contents of the `prevPtr` into your copy of the cache.
9. Go to step 4.

Something like:

```go
newPtr := &newCache
for !bldr.cache.CompareAndSwap(prevPtr, newPtr) {
  prevPtr = bldr.cache.Load()
  for k, v := range *prevPtr {
    newCache[k] = v
  }
}
```

Reply:
Ah right! Thats neat. It would duplicate the work if both `A` and `B` are building the same message but avoid blocking on different message types. I'll try it.

Reply:
yeah unfortunately the only way to block concurrent duplicate work is with a mutex

Reply:
Benchmarking is the same between the two, could try add more to clarify but might be overkill. Let me know if anything is needed or want to look at something else

Reply:
Okay, need to serialize writes to avoid a race on the `constaints.Cache`. I've got a benchmark with dynamic messages:
```go
type dynamicProto struct {
	proto.Message
	salt int32
}

func (d dynamicProto) ProtoReflect() protoreflect.Message {
	return dynamicMessage{d.Message.ProtoReflect(), d.salt}
}
//...
```
Each field is salted too. But I'll move it to an internal test.

Reply:
Race is on: 
- https://github.com/bufbuild/protovalidate-go/blob/e65d1395ab970830470fab8c928f4aadd3516c5d/internal/constraints/cache.go#L73
-  https://github.com/bufbuild/protovalidate-go/blob/e65d1395ab970830470fab8c928f4aadd3516c5d/internal/constraints/cache.go#L146